core/reorder/amd.cpp

```diff
@@ -0,0 +1,223 @@
+    IndexType* d_row_ptrs{};
+    IndexType* d_col_idxs{};
+    size_type d_nnz{};
+    if (auto convertible = dynamic_cast<const ConvertibleTo<complex_mtx>*>(
+            system_matrix.get())) {
+        auto conv_csr = complex_mtx::create(exec);
+        convertible->convert_to(conv_csr);
+        if (!parameters_.skip_sorting) {
+            conv_csr->sort_by_column_index();
+        }
+        if (!parameters_.skip_symmetrize) {
+            auto scalar =
+                initialize<complex_scalar>({one<std::complex<double>>()}, exec);
+            auto id = complex_identity::create(exec, conv_csr->get_size()[0]);
+            // compute A^T + A
+            conv_csr->transpose()->apply(scalar, id, scalar, conv_csr);
+        }
+        d_nnz = conv_csr->get_num_stored_elements();
+        d_row_ptrs = conv_csr->get_row_ptrs();
+        d_col_idxs = conv_csr->get_col_idxs();
+        converted = std::move(conv_csr);
+    } else {
+        auto conv_csr = real_mtx::create(exec);
+        as<ConvertibleTo<real_mtx>>(system_matrix)->convert_to(conv_csr);
+        if (!parameters_.skip_sorting) {
+            conv_csr->sort_by_column_index();
+        }
+        if (!parameters_.skip_symmetrize) {
+            auto scalar = initialize<real_scalar>({one<double>()}, exec);
+            auto id = real_identity::create(exec, conv_csr->get_size()[0]);
+            // compute A^T + A
+            conv_csr->transpose()->apply(scalar, id, scalar, conv_csr);
+        }
+        d_nnz = conv_csr->get_num_stored_elements();
+        d_row_ptrs = conv_csr->get_row_ptrs();
+        d_col_idxs = conv_csr->get_col_idxs();
+        converted = std::move(conv_csr);
+    }
+
+    // remove diagonal entries
+    auto pattern =
+        sparsity_mtx::create(exec, gko::dim<2>{num_rows, num_rows},
+                             make_array_view(exec, d_nnz, d_col_idxs),
+                             make_array_view(exec, num_rows + 1, d_row_ptrs));
+    pattern = pattern->to_adjacency_matrix();
+    // copy data to the CPU
+    array<IndexType> row_ptrs{host_exec, num_rows + 1};
+    host_exec->copy_from(exec, num_rows + 1, pattern->get_const_row_ptrs(),
+                         row_ptrs.get_data());
+    const auto nnz = row_ptrs.get_data()[num_rows];
+    array<IndexType> col_idxs_plus_workspace{host_exec,
+                                             nnz + nnz / 5 + 8 * num_rows};
+    host_exec->copy_from(exec, nnz, pattern->get_const_col_idxs(),
+                         col_idxs_plus_workspace.get_data());
+
+    array<IndexType> permutation{host_exec, num_rows};
+    array<IndexType> row_lengths{host_exec, num_rows};
+    for (size_type row = 0; row < num_rows; row++) {
+        row_lengths.get_data()[row] =
+            row_ptrs.get_data()[row + 1] - row_ptrs.get_data()[row];
+    }
+    // we use this much space for the column index workspace, the rest for
+    // row workspace
+    const auto col_idxs_plus_workspace_size = nnz + nnz / 5 + 2 * num_rows;
```

Comment:
maybe move this before allocating `col_idxs_plus_workspace`, with the size of (col_idxs_plus_workspace_size + 6 * num_rows)

Reply:
It only fits the suggestion of `iwlen : 1.2 * pfree +n` , right?